### PR TITLE
[WIP] [CR] Nested Item Containers

### PR DESCRIPTION
**Pull request by DavidKeaton**
# SUPER NOTE

I have begun working from scratch due to a fresh approach and merge conflicts that have accrued over time. I will be setting up a new PR for this. If you want to take a peek at the new branch, please refer to https://github.com/DavidKeaton/Cataclysm-DDA/tree/item_bags in the meantime. This branch aims to more or less refactor and rework the item class to make all the fun item stuff less of a hack approach and more of a correct one.

Thanks all!
-Davek
# END OF SUPER NOTE

**_Note!**_
This PR is in desperate need of some peer review, but I know ya'll been busy. _Whenever_ someone gets a chance, I'd love to make sure I haven't borked anything bad! :-)

_Items all up in Items!_
- contents.add() to add items.
- contents.rem() to remove items.
- contents.get() to vector of all contents.
- contents.has_\* functions to retrieve certain characteristics of contents [has_food, has_liquid, ...]
- various overloads to access it the same as before.

I have placed the storage class in item.h, to reduce clutter for fs and headers. `contents` is now class `storage` instead of a simple std::vector<item>, but is accessible _most_ of the same way as before (see above).

**_TODO**_
- [ ] Finish Container Filters (for autonomous item storage)
- [ ] Implement in `AIM`
- [ ] Better display for `inventory_ui`

Update 1:
- removed `self` pointer to item
- moved functions that determined space to item class

Update 2:
- reverted contents back to std::list<item>, not invstack (why I did this, God only knows)
- basic interface in inventory_ui (mostly for testing stuff atm)
- implementation of filter functions
- only allowing storage to containers that can store

Update 3:
- basic ui fixes in inventory for storing/removing (may still have some bugs) [for testing containers]
- _possible_ idea for filtering items based on certain criteria (to be expanded further)
- some other stuff I'm sure, can't remember (go diff checks!) :-)
- forgot to mention that I have added `push_back` and `erase` as per @BevapDin 

Update: Also as an aside, I am doing some smaller PRs to take a minor break from this so I don't burn out on it. Still in the works!

### CODE REVIEW

**Comment by DavidKeaton** [comment]
Correction, I think I have changes I am still doing, this may not compile ATM.

File: src/iuse_actor.cpp [hunk 1]
```diff
@@ -69,8 +69,8 @@ long iuse_transform::use(player *p, item *it, bool t, point /*pos*/) const
         // Transform into something in a container, assume the content is
         // "created" right now and give the content the current time as birthday
         it->make(container_id);
-        it->contents.push_back(item(target_id, calendar::turn));
-        target = &it->contents.back();
+        it->contents.add(item(target_id, calendar::turn));
+        target = &it->contents[it->contents.size()];
```
Comment (DavidKeaton) [inline src/iuse_actor.cpp]: I realized now this is redundant, doip.

Comment (KA101) [inline src/iuse_actor.cpp]: Side note: that comment up there looks Precarious if one stores perishables.  Careful with this one.

Comment (DavidKeaton) [inline src/iuse_actor.cpp]: Wait, wasn't this a bug in the current open issues?

Comment (KA101) [inline src/iuse_actor.cpp]: Possibly, but a quick search for "container" didn't turn up anything that looked relevant.

Comment (DavidKeaton) [inline src/iuse_actor.cpp]: noted, I will take a peek!

File: src/item.cpp [hunk 2]
```diff
@@ -5015,3 +5031,89 @@ bool item_category::operator!=( const item_category &rhs ) const
...
+
+/*-----------------------------------------------------------------------------
+ *  storage functions
+ *-----------------------------------------------------------------------------*/
+typedef std::vector<item>::iterator                                item_iterator;
+typedef std::vector<item>::const_iterator                    item_const_iterator;
+
+storage::storage(item *self, item_iterator start, item_iterator stop)
+{
+    me = self;
```
Comment (BevapDin) [inline src/item.cpp]: Linking back to the parent object is *usually* not a good approach. This implementation will give you crashes as the pointer is not updated by the copy constructor nor by the assignment operator.
```C++
item a = ... ; // a.content.me == &a
item b = ... ; // b.content.me == &b
a = b; // a.content.me == &b
```

File: src/item.h [hunk 3]
```diff
@@ -85,6 +88,96 @@ class item_category
...
+        // insert a set of items
+        std::vector<item *> add(std::vector<item> items, size_t index=0);
+        // insert a range of items
+        std::vector<item *> add(item_iterator start, item_iterator stop, size_t index=0);
+        // removes the item from storage, returning the item itself
+        item rem(size_t index);
+        // removes the item (from iterator)
+        item_iterator rem(item_iterator iter);
+        // removes a range of items
+        std::vector<item> rem(item_iterator start, item_iterator stop);
```
Comment (BevapDin) [inline src/item.h]: Why not use the `std::vector` function names `push_back` and `erase`, you don't have to update all the code that uses it.

File: src/item.h [hunk 4]
```diff
@@ -85,6 +88,96 @@ class item_category
...
+    typedef std::vector<item>::iterator         item_iterator;
+    typedef std::vector<item>::const_iterator   item_const_iterator;
+
+    private:
+        // <item *> to item that contains <this> storage instance
+        item *me;
+        // items stored in this bag
+        std::vector<item> items;
+        // holds any extra information we need about storage
+        std::map<std::string, void *> var;
```
Comment (BevapDin) [inline src/item.h]: And who manages the pointer stored here? Who deletes them?

File: src/item.cpp (hunk 2 above)
Comment (DavidKeaton) [inline src/item.cpp]: yeah I actually removed this altogether. lol.

File: src/item.h (hunk 3 above)
Comment (DavidKeaton) [inline src/item.h]: I had done just that, but wanted to make the names more sane for their purpose.

File: src/item.h (hunk 4 above)
Comment (DavidKeaton) [inline src/item.h]: it's a generic type, not solely a void pointer, used for type casted data. prob gonna remove.

**Comment by Zireael07** [comment]
@DavidKeaton: Any news on this?

**Comment by DavidKeaton** [comment]
tbh the aim bit took most of my time, but I want to finish this before the next release. :-)

**Comment by DavidKeaton** [comment]
I also have been considering other avenues for doing it, so it's not on the backburner completely! :-)

**Comment by DavidKeaton** [comment]
Please read the description as to why this is closed. (new branch)

**Pull request closed without merging**